doc/sphinx/addendum/type-classes.rst

```diff
@@ -295,10 +295,18 @@ the Existing Instance command to achieve the same effect.
 Summary of the commands
 -----------------------
 
-.. cmd:: Class @inductive_definition {* with @inductive_definition }
+.. cmd:: Class @record_definition
+         Class @class_definition
 
-   The :cmd:`Class` command is used to declare a typeclass with parameters
-   :n:`{* @binder }` and fields the declared record fields.
+   .. insertprodn class_definition class_definition
+
+   .. prodn::
+      class_definition ::= {? > } @ident_decl {* @binder } {? : @sort } {? := @constructor } {? @decl_notations }
```

Comment:
You don't need the optional brackets around the `:= @constructor` because this is already included in the `@record_definition` case.

Also, I suggest renaming `class_definition` into `singleton_class_definition`.

Finally, it looks like `@decl_notations` is not supported for singleton classes, nor even for `@record_definition`, despite what our documented syntax says!

Reply:
I think the brackets are necessary, otherwise how do you parse `Class foo.`, which is accepted?

> Also, I suggest renaming `class_definition` into `singleton_class_definition`.

Would you elaborate?  Singleton of what?  Seems like that would merit explanation in the text rather than an obscure name change for a nonterminal.

Also, would you help me complete the text (see "In the second form ...")?

Reply:
> Finally, it looks like `@decl_notations` is not supported for singleton classes, nor even for `@record_definition`, despite what our documented syntax says!

I would say it's a fail to have `Record` and `Class` based on `inductive_definition`.  WDYT about a PR to give them distinct grammar?

Reply:
> I think the brackets are necessary, otherwise how do you parse `Class foo.`, which is accepted?

This case is already included in the `record_definition` case (it is equivalent to `:= {}`).

>Would you elaborate? Singleton of what? Seems like that would merit explanation in the text rather than an obscure name change for a nonterminal.

It's how it's called in the refman: https://coq.inria.fr/refman/addendum/type-classes.html#singleton-class
It's also called "definitional classes". Maybe we should use this terminology everywhere instead...

> I would say it's a fail to have `Record` and `Class` based on `inductive_definition`. WDYT about a PR to give them distinct grammar?

I'd be happy to split the grammar a bit so that unsupported combinations (giving an inductive definition after a `Record` or `Structure` keyword) are excluded at parsing time. OTOH, `Inductive` and `CoInductive` must continue to support the record syntax (inductive and coinductive records). Finally, I don't think there is any justification for not supporting the `where` clause in `Record` and `Class`, so I'd say that ideally this limitation should be fixed.

Reply:
* What do you have against `Class foo.`?  It's a valid input and much, much shorter than `Class foo := {}.`  The user is just supposed to know to use the longer form?

* `singleton_class_definition` it is.

* I'll put in a PR to split out `Record` and `Class` separately, but I may need help to add `decl_notations` to them.

Reply:
>What do you have against `Class foo.`?  It's a valid input and much, much shorter than `Class foo := {}.`  The user is just supposed to know to use the longer form?

I have nothing against, I'm just saying this case is already taken care of in the `record_definition` syntax.

Reply:
OK, I was misreading the syntax when comparing `inductive_definition` to `record_definition` and `singleton_class_definition` (both of which are created by common.edit_mlg).

Reply:
> I'll put in a PR to split out Record and Class separately, but I may need help to add decl_notations to them.

I looked at this last night.  I won't do a PR on this.  There are almost 100 lines of validation code applied to `inductive_definition` in `Vernacentries.vernac_inductive`.  I see the commit comment from @ppedrot on parts of it "... Highly spaghetti code, hopefully works.".  There may be bugs lurking here.

Please see the todo I added in `records.rst`.